### PR TITLE
[mlir] Replace MLIR_ENABLE_CUDA_CONVERSIONS with LLVM_HAS_NVPTX_TARGET

### PR DESCRIPTION
**Pull request by tyb0807**
LLVM_HAS_NVPTX_TARGET is automatically set depending on whether NVPTX was enabled when building LLVM. Use this instead of manually defining MLIR_ENABLE_CUDA_CONVERSIONS (whose name is a bit misleading btw).

### CODE REVIEW

File: mlir/lib/Target/LLVM/NVVM/Target.cpp [hunk 1]
```diff
@@ -167,6 +167,15 @@ class NVPTXSerializer : public SerializeGPUModuleBase {
 
   gpu::GPUModuleOp getOperation();
 
+  std::optional<SmallVector<char, 0>>
```
Comment (ftynse) [inline mlir/lib/Target/LLVM/NVVM/Target.cpp:170]: I see that you are only moving this, but consider adding some documentation if you understand what this does.

File: mlir/lib/Target/LLVM/NVVM/Target.cpp [hunk 2]
```diff
@@ -167,6 +167,15 @@ class NVPTXSerializer : public SerializeGPUModuleBase {
...
+  moduleToObject(llvm::Module &llvmModule) override;
+
+private:
+  // Target options.
```
Comment (ftynse) [inline mlir/lib/Target/LLVM/NVVM/Target.cpp:174]:
```suggestion
  /// Target options.
```

File: mlir/lib/Target/LLVM/NVVM/Target.cpp [hunk 3]
```diff
@@ -549,22 +555,29 @@ NVPTXSerializer::moduleToObject(llvm::Module &llvmModule) {
...
     // Make sure to include the null terminator.
     StringRef bin(serializedISA->c_str(), serializedISA->size() + 1);
     return SmallVector<char, 0>(bin.begin(), bin.end());
   }
 
+  // At this point, compilation target is either `binary` or `fatbinary`, which
+  // requires CUDA toolkit.
+  if (!(MLIR_ENABLE_CUDA_CONVERSIONS)) {
+    getOperation().emitError(
+        "CUDA toolkit not provided when trying to serialize GPU module.");
```
Comment (ftynse) [inline mlir/lib/Target/LLVM/NVVM/Target.cpp:570]:
```suggestion
        "CUDA toolkit not provided when trying to serialize GPU module");
```

**Review by ftynse — APPROVED** [review]
Looks reasonable, IIRC, the design of the new GPU pipeline was intended to avoid a compile-time dependency on CUDA libraries.

Does the AMD backend suffer from the same issue?

File: mlir/lib/Target/LLVM/NVVM/Target.cpp (hunk 3 above)
Comment (grypp) [inline mlir/lib/Target/LLVM/NVVM/Target.cpp:570]: Can we test that path?

**Review by grypp — COMMENTED** [review]
Looks good, is it possible to add a test?

**Review by fabianmcg — CHANGES_REQUESTED** [review]
They're already decoupled, `MLIR_ENABLE_CUDA_CONVERSIONS` is an alias for checking whether the `NVPTX` target was built, see:
https://github.com/llvm/llvm-project/blob/main/mlir/CMakeLists.txt#L115C1-L119

Also, `moduleToObject` will succeed if `NVPTX` was built and the compilation target is `offload` or `ptx`, I know this because I use it frequently in a machine without the toolkit.

I'm ok with changing the name of the flag as it might be confusing.

**Comment by tyb0807** [comment]
> They're already decoupled, `MLIR_ENABLE_CUDA_CONVERSIONS` is an alias for checking whether the `NVPTX` target was built, see: https://github.com/llvm/llvm-project/blob/main/mlir/CMakeLists.txt#L115C1-L119
> 
> Also, `moduleToObject` will succeed if `NVPTX` was built and the compilation target is `offload` or `ptx`, I know this because I use it frequently in a machine without the toolkit.
> 
> I'm ok with changing the name of the flag as it might be confusing.

Yes, one of the changes introduced here is just replacing `MLIR_ENABLE_CUDA_CONVERSIONS` to `LLVM_HAS_NVPTX_TARGET`. However, I'm keeping `MLIR_ENABLE_CUDA_CONVERSIONS` flag to indicate whether CUDA toolkit is configured or not, as one can build LLVM with `NVPTX`, then accidentally try to serialize to `binary` without having the toolkit available. The part of the code that does actual serialization to binary (using `ptxas` or `nvptxcompiler`) will now be guarded by `MLIR_ENABLE_CUDA_CONVERSIONS` and should only be available if CUDA toolkit is configured.

Does this make sense?

**Comment by fabianmcg** [comment]
> The part of the code that does actual serialization to binary (using `ptxas` or `nvptxcompiler`) will now be guarded by `MLIR_ENABLE_CUDA_CONVERSIONS` and should only be available if CUDA toolkit is configured.
> 
> Does this make sense?

Not really. Because you're not changing the definition of `MLIR_ENABLE_CUDA_CONVERSIONS`, hence `MLIR_ENABLE_CUDA_CONVERSIONS` and `LLVM_HAS_NVPTX_TARGET` still fulfill the same purpose.

If the CUDA toolkit is not installed and try to serialize you'll get:
```
error: Couldn't find the `ptxas` binary. Please specify the toolkit path, add the compiler to $PATH, or set one of the environment variables in `NVVM::getCUDAToolkitPath()`.
```
And that's by design, as it allows to ship the functionality of compilation without requiring a hard dependency on the toolkit. This is also what `clang` does, you don't need the toolkit to build clang, only for compiling the code and if the toolkit is missing then fail gracefully.

**Comment by joker-eph** [comment]
Considering the discussion, the whole PR title and description will need a revamp I believe.

**Comment by tyb0807** [comment]
I've replaced `MLIR_ENABLE_CUDA_CONVERSIONS` with `LLVM_HAS_NVPTX_TARGET`. There are way more modifications than anticipated but these are NFC IMO.

**Comment by joker-eph** [comment]
I don't see a change in mlir-config.h ? Can you grep for MLIR_ENABLE_CUDA_CONVERSIONS and ensure it does not remain any occurrence in the codebase.

File: mlir/test/lit.site.cfg.py.in [hunk 4]
```diff
@@ -25,7 +25,7 @@ config.mlir_cmake_dir = "@MLIR_CMAKE_DIR@"
 config.mlir_lib_dir = "@MLIR_LIB_DIR@"
 
 config.build_examples = @LLVM_BUILD_EXAMPLES@
-config.run_cuda_tests = @MLIR_ENABLE_CUDA_CONVERSIONS@
+config.run_cuda_tests = "@LLVM_HAS_NVPTX_TARGET@"
```
Comment (joker-eph) [inline mlir/test/lit.site.cfg.py.in:28]: Can you rename `run_cuda_tests` as well? It is incorrectly named now.

**Comment by tyb0807** [comment]
>I don't see a change in mlir-config.h ? Can you grep for MLIR_ENABLE_CUDA_CONVERSIONS and ensure it does not remain any occurrence in the codebase
Indeed, I messed up during rebase somehow. Thanks!

**Comment by tyb0807** [comment]
@fabianmcg is it better now?

**Review by fabianmcg — APPROVED** [review]
LGTM, thank you! Just make sure that:
- all integration tests pass
- `ninja check-mlir` succeeds with and without the NVPTX being built.
- fix the description of the commit.

**Review by ftynse — CHANGES_REQUESTED** [review]
Looks even cleaner, thanks! But I suspect there is something off with how the macro is interpreted, it should be an always-defined 0/1 macro.

Note that Bazel is an optional build system for LLVM, you don't necessarily have to update it.

**Review by ftynse — APPROVED** [review]

File: mlir/lib/Target/LLVM/CMakeLists.txt [hunk 5]
```diff
@@ -47,7 +47,7 @@ add_mlir_dialect_library(MLIRNVVMTarget
   MLIRNVVMToLLVMIRTranslation
   )
 
-if(MLIR_ENABLE_CUDA_CONVERSIONS)
+if ("NVPTX" IN_LIST LLVM_TARGETS_TO_BUILD)
   # Find the CUDA toolkit.
   find_package(CUDAToolkit)
```
[the remark this answers is not in the document]
Comment (joker-eph) [reply mlir/lib/Target/LLVM/CMakeLists.txt:52]: This is marked resolved but I don't see the change?

**Pull request merged**